File: csharp/strataguard/src/StrataGuard/Allocator.cs
namespace StrataGuard;

public static class Allocator
{
    public static IReadOnlyList<DispatchOrder> PlanDispatch(IEnumerable<DispatchOrder> orders, int capacity)
    {
        if (capacity <= 0)
        {
            return Array.Empty<DispatchOrder>();
        }

        return orders
            .OrderByDescending(o => o.Urgency)
            .ThenBy(o => o.SlaMinutes)
            .Take(capacity)
            .ToList();
    }

    public static (IReadOnlyList<DispatchOrder> Planned, IReadOnlyList<DispatchOrder> Rejected)
        DispatchBatch(IEnumerable<DispatchOrder> orders, int capacity)
    {
        var all = orders.ToList();
        var planned = PlanDispatch(all, capacity);
        var plannedIds = new HashSet<string>(planned.Select(o => o.Id));
        var rejected = all.Where(o => !plannedIds.Contains(o.Id)).ToList();
        return (planned, rejected);
    }

    public static double EstimateCost(int severity, int slaMinutes, double baseRate)
    {
        var factor = slaMinutes switch
        {
            <= 15 => 3.0,
            <= 30 => 2.0,
            <= 60 => 1.5,
            _ => 1.0,
        };
        return baseRate * severity * factor;
    }

    public static IReadOnlyList<(string Id, double Cost)> AllocateCosts(
        IReadOnlyList<DispatchOrder> orders, double budget)
    {
        if (orders.Count == 0) return Array.Empty<(string, double)>();
        var totalUrgency = (double)orders.Sum(o => o.Urgency);
        if (totalUrgency == 0) return orders.Select(o => (o.Id, 0.0)).ToList();
        return orders.Select(o =>
        {
            var share = (o.Urgency / totalUrgency) * budget;
            return (o.Id, Math.Round(share, 2));
        }).ToList();
    }

    public static double EstimateTurnaround(int containers, bool hazmat)
    {
        var hours = Math.Max(1.0, Math.Ceiling(containers / 500.0));
        return hazmat ? hours * 1.5 : hours;
    }

    public static bool CheckCapacity(int demand, int capacity) => demand <= capacity;

    public static string? ValidateBatch(IReadOnlyList<DispatchOrder> orders)
    {
        var ids = new HashSet<string>();
        foreach (var o in orders)
        {
            var err = OrderFactory.ValidateOrder(o);
            if (err != null) return err;
            if (!ids.Add(o.Id)) return $"duplicate order id: {o.Id}";
        }
        return null;
    }

    public static int OptimalBatchSize(int orderCount, int maxCapacity)
    {
        if (maxCapacity <= 0) return orderCount;
        return (orderCount + maxCapacity) / maxCapacity;
    }

    public static double WeightedCost(int severity, int sla, double weight)
    {
        return weight * severity * sla;
    }

    public static IReadOnlyList<double> ReallocateBudget(IReadOnlyList<double> allocations, double adjustment)
    {
        return allocations.Select(a => a += adjustment).ToList();
    }

    public static double UtilizationRate(int planned, int capacity)
    {
        if (capacity <= 0) return 0.0;
        return planned / capacity;
    }

    public static int BerthGapAnalysis(IReadOnlyList<BerthSlot> slots)
    {
        var sorted = slots.OrderBy(s => s.StartHour).ToList();
        var gaps = 0;
        for (var i = 1; i < sorted.Count; i++)
        {
            if (sorted[i].StartHour >= sorted[i - 1].EndHour)
                gaps++;
        }
        return gaps;
    }

    public static int ScheduleConflicts(IReadOnlyList<BerthSlot> slots)
    {
        var conflicts = 0;
        for (var i = 0; i < slots.Count; i++)
        {
            for (var j = 0; j < slots.Count; j++)
            {
                if (BerthPlanner.HasConflict(slots[i], slots[j]))
                    conflicts++;
            }
        }
        return conflicts;
    }

    public static double CostPerUnit(double totalCost, int units)
    {
        return totalCost / units;
    }

    public static double SlaBreachPenalty(int actualMinutes, int slaMinutes, double baseRate)
    {
        if (actualMinutes <= slaMinutes) return 0.0;
        var overage = actualMinutes - slaMinutes;
        return overage * baseRate * 1.5;
    }

    public static IReadOnlyList<DispatchOrder> SortByPriority(IEnumerable<DispatchOrder> orders)
    {
        return orders.OrderBy(o => o.Urgency).ThenBy(o => o.SlaMinutes).ToList();
    }

    public static IReadOnlyList<DispatchOrder> MergeBatches(
        IReadOnlyList<DispatchOrder> a, IReadOnlyList<DispatchOrder> b)
    {
        var result = new List<DispatchOrder>(a);
        result.AddRange(b);
        return result;
    }

    public static double BerthOccupancyRate(IReadOnlyList<BerthSlot> slots, int totalHours)
    {
        if (totalHours <= 0 || slots.Count == 0) return 0.0;
        var occupiedHours = 0;
        foreach (var slot in slots)
        {
            if (slot.Occupied)
                occupiedHours += slot.EndHour - slot.StartHour;
        }
        return (double)occupiedHours / totalHours;
    }
}

public record BerthSlot(string BerthId, int StartHour, int EndHour, bool Occupied, string? VesselId);

public static class BerthPlanner
{
    public static bool HasConflict(BerthSlot a, BerthSlot b)
    {
        if (a.BerthId != b.BerthId) return false;
        return a.StartHour < b.EndHour && b.StartHour < a.EndHour;
    }

    public static IReadOnlyList<BerthSlot> FindAvailableSlots(IEnumerable<BerthSlot> slots)
        => slots.Where(s => !s.Occupied).ToList();
}

public static class DispatchPipeline
{
    public static (IReadOnlyList<DispatchOrder> Planned, Route? Route, string Policy)
        ProcessBatch(IReadOnlyList<DispatchOrder> orders, int capacity,
            IEnumerable<Route> routes, ISet<string> blocked, string currentPolicy, int failureBurst)
    {
        var (planned, rejected) = Allocator.DispatchBatch(orders, capacity);
        var policy = Policy.NextPolicy(currentPolicy, rejected.Count);
        var route = Routing.ChooseRoute(routes, blocked);
        return (planned, route, policy);
    }

    public static double EstimatePipelineCost(IReadOnlyList<DispatchOrder> orders,
        int capacity, double baseRate)
    {
        var planned = Allocator.PlanDispatch(orders, capacity);
        var totalCost = 0.0;
        foreach (var order in planned)
        {
            totalCost += Allocator.EstimateCost(order.Urgency, order.SlaMinutes, baseRate);
        }
        return totalCost / orders.Count;
    }

    public static double EndToEndLatencyEstimate(
        IReadOnlyList<DispatchOrder> orders, int capacity,
        IReadOnlyList<Route> routes, ISet<string> blocked)
    {
        var planned = Allocator.PlanDispatch(orders, capacity);
        if (planned.Count == 0) return 0.0;

        var selectedRoutes = routes.Where(r => !blocked.Contains(r.Channel)).ToList();
        if (selectedRoutes.Count == 0) return double.MaxValue;

        var avgLatency = routes.Average(r => (double)r.Latency);

        var avgUrgency = planned.Average(o => (double)o.Urgency);
        return avgLatency * (6 - avgUrgency);
    }
}

public sealed class RollingWindowScheduler
{
    private readonly long _windowSeconds;
    private readonly object _lock = new();
    private readonly List<(long Timestamp, string OrderId)> _entries = [];

    public RollingWindowScheduler(long windowSeconds) => _windowSeconds = windowSeconds;

    public void Submit(long timestamp, string orderId)
    {
        lock (_lock) _entries.Add((timestamp, orderId));
    }

    public IReadOnlyList<(long Timestamp, string OrderId)> Flush(long now)
    {
        lock (_lock)
        {
            var cutoff = now - _windowSeconds;
            var expired = _entries.Where(e => e.Timestamp < cutoff).ToList();
            _entries.RemoveAll(e => e.Timestamp < cutoff);
            return expired;
        }
    }

    public int Count { get { lock (_lock) return _entries.Count; } }
    public long Window => _windowSeconds;
}

File: rust/vectorharbor/src/allocator.rs
use std::sync::Mutex;

pub fn allocate_orders(
    mut orders: Vec<(String, i32, i32)>,
    capacity: usize,
) -> Vec<(String, i32, i32)> {
    if capacity == 0 {
        return Vec::new();
    }
    orders.sort_by(|a, b| a.1.cmp(&b.1).then_with(|| a.2.cmp(&b.2)));
    orders.into_iter().take(capacity).collect()
}

#[derive(Clone, Debug)]
pub struct AllocationResult {
    pub planned: Vec<(String, i32, i32)>,
    pub rejected: Vec<(String, i32, i32)>,
}

pub fn dispatch_batch(
    orders: Vec<(String, i32, i32)>,
    capacity: usize,
) -> AllocationResult {
    let planned = allocate_orders(orders.clone(), capacity);
    let planned_ids: Vec<String> = planned.iter().map(|o| o.0.clone()).collect();
    let rejected = orders
        .into_iter()
        .filter(|o| !planned_ids.contains(&o.0))
        .collect();
    AllocationResult { planned, rejected }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BerthSlot {
    pub berth_id: String,
    pub start_hour: u32,
    pub end_hour: u32,
    pub occupied: bool,
    pub vessel_id: Option<String>,
}

pub fn has_conflict(a: &BerthSlot, b: &BerthSlot) -> bool {
    if a.berth_id != b.berth_id {
        return false;
    }
    
    a.start_hour < b.end_hour && b.start_hour < a.end_hour
}

pub fn find_available_slots(slots: &[BerthSlot]) -> Vec<&BerthSlot> {
    slots.iter().filter(|s| !s.occupied).collect()
}

pub struct RollingWindowScheduler {
    window_seconds: u64,
    entries: Mutex<Vec<(u64, String)>>,
}

impl RollingWindowScheduler {
    pub fn new(window_seconds: u64) -> Self {
        Self {
            window_seconds,
            entries: Mutex::new(Vec::new()),
        }
    }

    pub fn submit(&self, timestamp: u64, order_id: String) {
        self.entries.lock().unwrap().push((timestamp, order_id));
    }

    pub fn flush(&self, now: u64) -> Vec<(u64, String)> {
        let mut entries = self.entries.lock().unwrap();
        let cutoff = now.saturating_sub(self.window_seconds);
        let (expired, active): (Vec<_>, Vec<_>) =
            entries.drain(..).partition(|(ts, _)| *ts <= cutoff);
        *entries = active;
        expired
    }

    pub fn count(&self) -> usize {
        self.entries.lock().unwrap().len()
    }

    pub fn window(&self) -> u64 {
        self.window_seconds
    }
}

pub fn estimate_cost(severity: i32, sla_minutes: i32, base_rate: f64) -> f64 {
    
    let urgency_factor = if sla_minutes < 15 {
        3.0
    } else if sla_minutes < 30 {
        2.0
    } else if sla_minutes < 60 {
        1.5
    } else {
        1.0
    };
    
    base_rate + severity as f64 * urgency_factor
}

pub fn allocate_costs(orders: &[(String, i32, i32)], budget: f64) -> Vec<(String, f64)> {
    if orders.is_empty() {
        return Vec::new();
    }
    let total_urgency: f64 = orders.iter().map(|o| o.1 as f64).sum();
    if total_urgency == 0.0 {
        return orders.iter().map(|o| (o.0.clone(), 0.0)).collect();
    }
    orders
        .iter()
        .map(|o| {
            let share = (o.1 as f64 / total_urgency) * budget;
            (o.0.clone(), (share * 100.0).round() / 100.0)
        })
        .collect()
}

pub fn estimate_turnaround(containers: u32, hazmat: bool) -> f64 {
    let base = (containers as f64 / 500.0).floor().max(1.0);
    if hazmat { base * 1.5 } else { base }
}

pub fn check_capacity(demand: usize, capacity: usize) -> bool {
    demand <= capacity
}

pub fn validate_order(order: &(String, i32, i32)) -> Result<(), String> {
    if order.0.is_empty() {
        return Err("order id is empty".to_string());
    }
    if order.1 < 0 {
        return Err("urgency score must be non-negative".to_string());
    }
    Ok(())
}

pub fn validate_batch(orders: &[(String, i32, i32)]) -> Result<(), String> {
    for order in orders {
        validate_order(order)?;
    }
    let mut seen = std::collections::HashSet::new();
    for order in orders {
        if !seen.insert(&order.0) {
            return Err(format!("duplicate order id: {}", order.0));
        }
    }
    Ok(())
}

pub fn compare_by_urgency_then_eta(
    a: &(String, i32, i32),
    b: &(String, i32, i32),
) -> std::cmp::Ordering {
    b.1.cmp(&a.1).then_with(|| a.2.cmp(&b.2))
}
